Libraries/D3D12RaytracingFallback/src/TreeletReorderBindings.h

```diff
@@ -28,22 +28,50 @@ struct InputConstants
 #define NumTrianglesBufferRegister 1
 #define AABBBufferRegister 2
 #define ElementBufferRegister 3
+#define BaseTreeletsCountBufferRegister 4
+#define BaseTreeletsIndexBufferRegister 5
 
 #define GlobalDescriptorHeapRegister 0
 #define GlobalDescriptorHeapRegisterSpace 1
 
+static const uint FullTreeletSize = 7;
+
 #ifdef HLSL
 // These need to be UAVs despite being read-only because the fallback layer only gets a 
 // GPU VA and the API doesn't allow any way to transition that GPU VA from UAV->SRV
+RWStructuredBuffer<Primitive> InputBuffer : UAV_REGISTER(ElementBufferRegister);
 
+globallycoherent RWByteAddressBuffer NumTrianglesBuffer : UAV_REGISTER(NumTrianglesBufferRegister);
 globallycoherent RWStructuredBuffer<HierarchyNode> hierarchyBuffer : UAV_REGISTER(HierarchyBufferRegister);
-RWByteAddressBuffer NumTrianglesBuffer : UAV_REGISTER(NumTrianglesBufferRegister);
 globallycoherent RWStructuredBuffer<AABB> AABBBuffer : UAV_REGISTER(AABBBufferRegister);
-RWStructuredBuffer<Primitive> InputBuffer : UAV_REGISTER(ElementBufferRegister);
-RWByteAddressBuffer DescriptorHeapBufferTable[] : UAV_REGISTER_SPACE(GlobalDescriptorHeapRegister, GlobalDescriptorHeapRegisterSpace);
+
+globallycoherent RWByteAddressBuffer BaseTreeletsCountBuffer : UAV_REGISTER(BaseTreeletsCountBufferRegister);
+RWStructuredBuffer<uint> BaseTreeletsIndexBuffer : UAV_REGISTER(BaseTreeletsIndexBufferRegister);
 
 cbuffer TreeletConstants : CONSTANT_REGISTER(ConstantsRegister)
 {
     InputConstants Constants;
 };
-#endif
+
+static const uint NumInternalTreeletNodes = FullTreeletSize - 1;
+
+static const uint NumTreeletSplitPermutations = 1 << FullTreeletSize;
+static const uint FullPartitionMask = NumTreeletSplitPermutations - 1;
+
+static const uint RootNodeIndex = 0; 
+
+// All results of FullTreeletSize choose i 
+static const uint FullTreeletSizeChoose[FullTreeletSize + 1] = { 1, 7, 21, 35, 35, 21, 7, 1 };
+
+// BitPermutations[k][n] = the n'th permutation of FullTreeletSize bits with k bits set
+static const uint BitPermutations[FullTreeletSize + 1][35] = {
+    { 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00, 0x00 },
```

Comment:
It doesn't seem like the first 2 rows ever get used. In theory if the loop is actually getting unrolled, the compiler should be smart enough to figure this out, I'd be tempted to take a peek at the DXIL and confirm that's the case, otherwise this is just wasted memory going down the drain. (Feel free to grab me for this)

Reply:
This was a fairly substantial perf win